### PR TITLE
Add support for Emergency Access

### PR DESCRIPTION
**Pull request by Hinton**
Adds support for the new Emergency Access feature. In short it provides a way to designate an emergency contact whom after a configurable timeout may either takeover the account or be granted view access.

https://app.asana.com/0/1195018406457675/1153292148278672

Relevant PRs:
https://github.com/bitwarden/jslib/pull/204
https://github.com/bitwarden/server/pull/1000

<details>
<summary>Screenshots</summary>

### View contacts
image

### Edit
image

### Request access
image

### Approve / Reject
image

### View users vault
image

image

### Takeover

image

</details>

### CODE REVIEW

File: src/app/settings/emergency-access-takeover.component.html [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            <div class="modal-header">
+                <h2 class="modal-title" id="userAddEditTitle">
+                    {{'takeover' | i18n}}
+                    <small class="text-muted" *ngIf="name">{{name}}</small>
+                </h2>
+                <button type="button" class="close" data-dismiss="modal" appA11yTitle="{{'close' | i18n}}">
+                    <span aria-hidden="true">&times;</span>
+                </button>
+            </div>
+
```
Comment (addisonbeck) [inline src/app/settings/emergency-access-takeover.component.html:13]:
```suggestion
```

File: src/app/settings/emergency-access-takeover.component.html [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                    <small class="text-muted" *ngIf="name">{{name}}</small>
+                </h2>
+                <button type="button" class="close" data-dismiss="modal" appA11yTitle="{{'close' | i18n}}">
+                    <span aria-hidden="true">&times;</span>
+                </button>
+            </div>
+            
+            <div class="modal-body">
+                <app-callout type="warning">{{'loggedOutWarning' | i18n}}</app-callout>
+
```
Comment (addisonbeck) [inline src/app/settings/emergency-access-takeover.component.html:16]:
```suggestion
```

File: src/app/settings/emergency-access-takeover.component.html [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                        <div class="form-group">
+                            <label for="masterPasswordRetype">{{'confirmNewMasterPass' | i18n}}</label>
+                            <input id="masterPasswordRetype" type="password" name="MasterPasswordRetype"
+                                class="form-control" [(ngModel)]="masterPasswordRetype" required appInputVerbatim
+                                autocomplete="new-password">
+                        </div>
+                    </div>
+                </div>
+            </div>
+
```
Comment (addisonbeck) [inline src/app/settings/emergency-access-takeover.component.html:37]:
```suggestion
```

File: src/app/settings/emergency-access-takeover.component.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    @Input() kdf: KdfType;
+    @Input() kdfIterations: number;
+
+    formPromise: Promise<any>;
+
+    constructor(i18nService: I18nService,
+        cryptoService: CryptoService, messagingService: MessagingService,
+        userService: UserService, passwordGenerationService: PasswordGenerationService,
+        platformUtilsService: PlatformUtilsService, policyService: PolicyService,
+        private apiService: ApiService, private toasterService: ToasterService) {
```
Comment (addisonbeck) [inline src/app/settings/emergency-access-takeover.component.ts:43]:
```suggestion
    constructor(i18nService: I18nService, cryptoService: CryptoService,
        messagingService: MessagingService, userService: UserService,
        passwordGenerationService: PasswordGenerationService,
        platformUtilsService: PlatformUtilsService, policyService: PolicyService,
        private apiService: ApiService, private toasterService: ToasterService) {
```

File: src/app/settings/emergency-access-view.component.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    Component,
+    ComponentFactoryResolver,
+    OnInit,
+    ViewChild,
+    ViewContainerRef,
+} from '@angular/core';
+import { ActivatedRoute, Router } from '@angular/router';
+import { ApiService, CryptoService } from 'jslib/abstractions';
+
+import { CipherService } from 'jslib/abstractions/cipher.service';
```
Comment (addisonbeck) [inline src/app/settings/emergency-access-view.component.ts:11]:
```suggestion
import { ApiService } from 'jslib/abstractions/api.service';
import { CipherService } from 'jslib/abstractions/cipher.service';
import { CryptoService } from 'jslib/abstractions/crypto.service';
```

File: src/app/settings/emergency-access-view.component.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import { CipherData } from 'jslib/models/data';
+import { Cipher, SymmetricCryptoKey } from 'jslib/models/domain';
+import { EmergencyAccessViewResponse } from 'jslib/models/response/emergencyAccessResponse';
+
+import { CipherView } from 'jslib/models/view/cipherView';
```
Comment (addisonbeck) [inline src/app/settings/emergency-access-view.component.ts:16]:
```suggestion
import { CipherView } from 'jslib/models/view/cipherView';

```

File: src/locales/en/messages.json [hunk 7]
```diff
@@ -3368,6 +3371,139 @@
...
+    "message": "Invite a new emergency contact by entering their Bitwarden account email address below. If they do not have a Bitwarden account already, they will be prompted to create a new account."
+  },
+  "emergencyAccessRecoveryInitiated": {
+    "message": "Emergency Access Initiated"
+  },
+  "emergencyAccessRecoveryApproved": {
+    "message": "Emergency Access Approved"
+  },
+  "viewDesc": {
+    "message": "Can view all entries in your own vault."
```
Comment (cscharf) [inline src/locales/en/messages.json:3405]: `entries` should be `items`

File: src/app/settings/emergency-access.component.html [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    </h2>
+    <div class="ml-auto d-flex">
+        <button class="btn btn-sm btn-outline-primary ml-3" type="button" (click)="invite()" [disabled]="!canAccessPremium">
+            <i aria-hidden="true" class="fa fa-plus fa-fw"></i>
+            {{'addEmergencyContact' |i18n}}
+        </button>
+    </div>
+</div>
+
+<table class="table table-hover table-list mb-0" *ngIf="trustedContacts && trustedContacts.length">
```
Comment (cscharf) [inline src/app/settings/emergency-access.component.html:21]: Need a "no items" message when there are no items to add to the table.

File: src/app/settings/emergency-access.component.html [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            </td>
+        </tr>
+    </tbody>
+</table>
+
+<div class="page-header spaced-header">
+    <h2>{{'desegnatedEmergencyContacts' | i18n}}</h2>
+</div>
+
+<table class="table table-hover table-list mb-0" *ngIf="grantedContacts && grantedContacts.length">
```
Comment (cscharf) [inline src/app/settings/emergency-access.component.html:89]: Same, need a "no items" message when there are no items to add to the table.

File: src/app/settings/emergency-access-add-edit.component.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    waitTimes: { name: string; value: number; }[];
+    waitTime: number;
+
+    constructor(private apiService: ApiService, private i18nService: I18nService,
+        private toasterService: ToasterService) { }
+
+    async ngOnInit() {
+        this.editMode = this.loading = this.emergencyAccessId != null;
+
+        this.waitTimes = [
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/app/settings/emergency-access-add-edit.component.ts:47]: They are currently hard coded as a dropdown list. But there's nothing preventing us from changing it to a text field. However I feel that giving to many options could be disadvantageous.

File: src/app/settings/emergency-access.component.ts [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        private storageService: StorageService, private userService: UserService,
+        private messagingService: MessagingService) { }
+
+    async ngOnInit() {
+        this.canAccessPremium = await this.userService.canAccessPremium();
+        this.load();
+    }
+
+    async load() {
+        this.trustedContacts = (await this.apiService.getEmergencyAccessTrusted()).data;
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/app/settings/emergency-access.component.ts:55]: They shouldn't the server should provide a `ListResponse`, and we don't have those checks in other places.

File: src/app/settings/emergency-access.component.ts [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+        if (this.modal != null) {
+            this.modal.close();
+        }
+
+        const factory = this.componentFactoryResolver.resolveComponentFactory(ModalComponent);
+        this.modal = this.addEditModalRef.createComponent(factory).instance;
+        const childComponent = this.modal.show<EmergencyAccessAddEditComponent>(
+            EmergencyAccessAddEditComponent, this.addEditModalRef);
+
+        childComponent.name = details != null ? details.name || details.email : null;
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/app/settings/emergency-access.component.ts:76]: Hmm nice, "Nullish Coalescing". Didn't know they had added that to TS.

**Review by cscharf — APPROVED** [review]

**Pull request merged**